### PR TITLE
[`Mixtral`] Fix loss + nits

### LINKED ISSUE #28021: Incorrect router probability calculation
### System Info

transformers version 4.36.0

### Who can help?

 @ArthurZucker and @younesbelkada

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I think load_balancing_loss_func in modeling_mixtral creates router_prob_per_group_and_expert incorrectly
https://github.com/huggingface/transformers/blob/v4.36.0/src/transformers/models/mixtral/modeling_mixtral.py#L120

Trying to multiply something batch_size * num_hidden_layers, num_experts by batch_size * num_hidden_layers, topk, 1

`torch.mean(tokens_per_group_and_expert * router_prob_per_group_and_expert.unsqueeze(-1)) * (num_experts**2)`

Correct creation of routing_weights should likely be from gate_logits, which ensures it is the correct size

`routing_weights = gate_logits.softamx(dim=-1)`

The unsqueeze(-1) is necessary with this. Also router_prob_per_group_and_expert should average over axis=-2

`router_prob_per_group_and_expert = torch.mean(routing_weights, axis=-2)`

This follows the previous implementation in modeling_switch_transformers
https://github.com/huggingface/transformers/blob/v4.36.0/src/transformers/models/switch_transformers/modeling_switch_transformers.py#L91

### Expected behavior

Something like this would fix it

```
def router_loss_func_test(gate_logits: torch.Tensor, top_k=2) -> float:
    if gate_logits is None:
        return 0

    if isinstance(gate_logits, tuple):
        # cat along the layers?
        gate_logits = torch.cat(gate_logits, dim=0) # batch_size * num_hidden_layers, sequence_length, num_experts

    num_experts = gate_logits.shape[-1]

    _, expert_indicies = torch.topk(gate_logits, top_k, dim=-1)  # this is done so you don't need to pass expert_indicies
    routing_probs = gate_logits.softmax(dim=-1) # routing probs

    if expert_indicies.dtype != torch.int64: # cast the expert indices to int64, otherwise one-hot encoding will fail
        expert_indicies = expert_indicies.to(torch.int64)

    if len(expert_indicies.shape) == 2:
        expert_indicies = expert_indicies.unsqueeze(2)

    expert_mask = torch.nn.functional.one_hot(expert_indicies, num_experts)

    # For a given token, determine if it was routed to a given expert.
    expert_mask = torch.max(expert_mask, axis=-2).values

    expert_mask = expert_mask.to(torch.float32) # cast to float32 otherwise mean will fail
    tokens_per_group_and_expert = torch.mean(expert_mask, axis=-2)

    router_prob_per_group_and_expert = torch.mean(routing_probs, axis=-2)

    loss = torch.mean(tokens_per_group_and_expert * router_prob_per_group_and_expert) * (num_experts**2)
    return loss
```

### LINKED ISSUE #28093: load_balancing_loss in mixtral model
### System Info

torch '1.13.0+cu117'

### Who can help?

@ArthurZucker and @younesbelkada

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The balancing loss function always return a constant.
Here is the official code:
```
def load_balancing_loss_func(gate_logits: torch.Tensor, num_experts: torch.Tensor = None, top_k=2) -> float:
    r"""
    Computes auxiliary load balancing loss as in Switch Transformer - implemented in Pytorch.

    See Switch Transformer (https://arxiv.org/abs/2101.03961) for more details. This function implements the loss
    function presented in equations (4) - (6) of the paper. It aims at penalizing cases where the routing between
    experts is too unbalanced.

    Args:
        gate_logits (Union[`torch.Tensor`, Tuple[torch.Tensor]):
            Logits from the `gate`, should be a tuple of tensors. Shape: [batch_size, seqeunce_length, num_experts].
        num_experts (`int`, *optional*):
            Number of experts

    Returns:
        The auxiliary loss.
    """
    if gate_logits is None:
        return 0

    if isinstance(gate_logits, tuple):
        # cat along the layers?
        gate_logits = torch.cat(gate_logits, dim=0)

    routing_weights, selected_experts = torch.topk(gate_logits, top_k, dim=-1)
    routing_weights = routing_weights.softmax(dim=-1)

    # cast the expert indices to int64, otherwise one-hot encoding will fail
    if selected_experts.dtype != torch.int64:
        selected_experts = selected_experts.to(torch.int64)

    if len(selected_experts.shape) == 2:
        selected_experts = selected_experts.unsqueeze(2)

    expert_mask = torch.nn.functional.one_hot(selected_experts, num_experts)

    # For a given token, determine if it was routed to a given expert.
    expert_mask = torch.max(expert_mask, axis=-2).values

    # cast to float32 otherwise mean will fail
    expert_mask = expert_mask.to(torch.float32)
    tokens_per_group_and_expert = torch.mean(expert_mask, axis=-2)

    router_prob_per_group_and_expert = torch.mean(routing_weights, axis=-1)
    return torch.mean(tokens_per_group_and_expert * router_prob_per_group_and_expert.unsqueeze(-1)) * (num_experts**2)
```
Here is my code:
```
num_hidden_layers=30
batch_size = 16
seq_len = 32
num_experts = 8
gate_logits = tuple(torch.randn(batch_size*seq_len, num_experts) for _ in range(num_hidden_layers))
load_balancing_loss_func(gate_logits=gate_logits, num_experts=num_experts)
```
It always return 4.

### Expected behavior

please anwser this question

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Properly compute the loss. Pushes for a uniform distribution. 

fixes #28021
Fixes https://github.com/huggingface/transformers/issues/28093

### CODE REVIEW

File: src/transformers/tokenization_utils_fast.py [hunk 1]
```diff
@@ -500,7 +500,7 @@ def _batch_encode_plus(
             stride=stride,
             pad_to_multiple_of=pad_to_multiple_of,
         )
-
+        print(batch_text_or_text_pairs)
```
Comment (younesbelkada) [inline src/transformers/tokenization_utils_fast.py:503]:
```suggestion
```

File: src/transformers/models/mixtral/configuration_mixtral.py [hunk 2]
```diff
@@ -128,7 +128,7 @@ def __init__(
         eos_token_id=2,
         tie_word_embeddings=False,
         rope_theta=1e6,
-        sliding_window=4096,
+        sliding_window=None,
```
Comment (younesbelkada) [inline src/transformers/models/mixtral/configuration_mixtral.py:131]: should we completely remove all logic with respect to `sliding_window` in modeling mixtral?

**Review by younesbelkada — COMMENTED** [review]
Thanks very much, the changes look great, I left one open question about the modeling file

File: src/transformers/tokenization_utils_fast.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/tokenization_utils_fast.py:503]: lol

File: src/transformers/models/mixtral/configuration_mixtral.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/mixtral/configuration_mixtral.py:131]: It's not BC

Comment (ArthurZucker) [reply src/transformers/models/mixtral/configuration_mixtral.py:131]: So no we can't

Comment (younesbelkada) [reply src/transformers/models/mixtral/configuration_mixtral.py:131]: In any case the mixtral team has already removed the `sliding_window` attribute in the config and it is already breaking as the argument is silently getting ignored in the modeling file: https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_attn_mask_utils.py#L149 / https://github.com/huggingface/transformers/blob/main/src/transformers/models/mixtral/modeling_mixtral.py#L1063

Comment (younesbelkada) [reply src/transformers/models/mixtral/configuration_mixtral.py:131]: We can maybe deprecate the arg through a deprecation cycle

Comment (younesbelkada) [reply src/transformers/models/mixtral/configuration_mixtral.py:131]: As discussed offline, let's keep it for performance BC for users that have pushed a mixtral model with sliding window

File: tests/models/mixtral/test_modeling_mixtral.py [hunk 3]
```diff
@@ -469,15 +469,16 @@ def test_load_balancing_loss(self):
...
         input_ids = input_dict["input_ids"]
         attention_mask = input_ids.ne(1).to(torch_device)
         model = MixtralForCausalLM(config)
         model.to(torch_device)
         model.eval()
         result = model(input_ids, attention_mask=attention_mask)
-        self.assertEqual(result.router_logits[0].shape, (91, config.num_experts_per_tok))
-        torch.testing.assert_close(result.aux_loss.cpu(), torch.tensor(1, dtype=torch.float32))
+        self.assertEqual(result.router_logits[0].shape, (91, config.num_local_experts))
+        torch.testing.assert_close(result.aux_loss.cpu(), torch.tensor(7.999999523162842, dtype=torch.float32))
```
Comment (LysandreJik) [inline tests/models/mixtral/test_modeling_mixtral.py:481]:
```suggestion
        torch.testing.assert_close(result.aux_loss.cpu(), torch.tensor(8, dtype=torch.float32))
```

**Review by LysandreJik — APPROVED** [review]
Thanks for the offline explanation, LGTM

**Review by younesbelkada — APPROVED** [review]
Thanks! let's 🚢 it

**Comment by teknium1** [comment]
> # What does this PR do?
> Properly compute the loss. Pushes for a uniform distribution.
> 
> fixes #28021 Fixes #28093

What were the side effects of the issue? Did it actually degrade training runs
